### PR TITLE
Implement RTDS

### PR DESCRIPTION
**Pull request by ivanhan**
Implement the RTDS API on server.go. I've added more comments in the changes below. 

Ran unit tests and integration tests successfully.

Thanks!

### CODE REVIEW

File: pkg/test/resource/resource.go [hunk 1]
```diff
@@ -367,11 +380,18 @@ func (ts TestSnapshot) Generate() cache.Snapshot {
 		listeners[i] = listener
 	}
 
+	runtimes := make([]cache.Resource, ts.NumRuntimes)
+	for i := 0; i < ts.NumRuntimes; i++ {
+		name := fmt.Sprintf("runtime-%d", i)
```
Comment (ivanhan) [inline pkg/test/resource/resource.go]: Not sure if this is an appropriate ask, but can RTDS resources be fully described by the API? Currently, the name of the RTDS runtime layers exists in two places: in the RTDS resource returned by the RTDS API here, and in the static  config file (see change below). This creates a coupling between the RTDS API and the config file. Hope I'm making sense :)

File: sample/bootstrap-xds.yaml [hunk 2]
```diff
@@ -32,3 +32,14 @@ static_resources:
         port_value: 18000
     http2_protocol_options: {}
     name: xds_cluster
+layered_runtime:
+  layers:
+    - name: runtime-0
```
Comment (ivanhan) [inline sample/bootstrap-xds.yaml]: Can the RTDS API describe the layers fully without having to define the layers in the config file as well?

File: pkg/cache/resource.go [hunk 3]
```diff
@@ -32,12 +33,14 @@ type Resource interface {
...
 // Resource types in xDS v2.
 const (
-	typePrefix   = "type.googleapis.com/envoy.api.v2."
-	EndpointType = typePrefix + "ClusterLoadAssignment"
-	ClusterType  = typePrefix + "Cluster"
-	RouteType    = typePrefix + "RouteConfiguration"
-	ListenerType = typePrefix + "Listener"
-	SecretType   = typePrefix + "auth.Secret"
+	apiTypePrefix   = "type.googleapis.com/envoy.api.v2."
+	discoTypePrefix = "type.googleapis.com/envoy.service.discovery.v2."
```
Comment (ivanhan) [inline pkg/cache/resource.go]: Minor renaming of the prefix

File: pkg/cache/resource.go [hunk 4]
```diff
@@ -51,6 +54,7 @@ var (
 		RouteType,
 		ListenerType,
 		SecretType,
+		RuntimeType,
```
Comment (ivanhan) [inline pkg/cache/resource.go]: Most of the changes to support RTDS just mimic what's been done for EDS, LDS, RDS, etc.

File: pkg/test/resource/resource.go [hunk 5]
```diff
@@ -294,6 +297,14 @@ func MakeTCPListener(listenerName string, port uint32, clusterName string) *v2.L
 	}
 }
 
+// MakeRuntime creates an RTDS layer with no fields.
+func MakeRuntime(runtimeName string) *discovery.Runtime {
+	return &discovery.Runtime{
+		Name:  runtimeName,
+		Layer: &pstruct.Struct{},
```
Comment (ivanhan) [inline pkg/test/resource/resource.go]: `MakeRuntime` currently just returns an RTDS layer with 0 fields defined.

**Comment by mattklein123** [comment]
cc @kyessenov @yangminzhu PTAL

**Comment by ivanhan** [comment]
@yangminzhu definitely! i'll add a field to that RTDS layer. will also update the README

P.S. anyone know how I can get access to reply to comments, hahaha :D

File: pkg/cache/resource.go (hunk 3 above)
Comment (kyessenov) [reply pkg/cache/resource.go]: nit: discovery vs disco 💃 🕺

File: sample/bootstrap-xds.yaml (hunk 2 above)
Comment (kyessenov) [reply sample/bootstrap-xds.yaml]: Does RTDS support collections of layers? E.g. can you retrieve all layers without naming them?

**Comment by ivanhan** [comment]
@kyessenov to confirm, I should exercise RTDS in ADS and REST as well? thanks for the reviews all!

File: pkg/cache/resource.go (hunk 3 above)
Comment (ivanhan) [reply pkg/cache/resource.go]: you bet!

File: sample/bootstrap-xds.yaml (hunk 2 above)
Comment (ivanhan) [reply sample/bootstrap-xds.yaml]: i know RTDS supports multiple layers, but I don't think the API itself supports retrieval of more than 1 layer at a time. will need to double check :)

**Comment by kyessenov** [comment]
Yes, I'd like to see ADS one. REST is optional.

**Comment by ivanhan** [comment]
roger that!

File: sample/bootstrap-xds.yaml (hunk 2 above)
Comment (ivanhan) [reply sample/bootstrap-xds.yaml]: The API might only support retrieving single layers at a time. E.g. when I specify 2 RTDS layers in my Envoy config file:
```
layered_runtime:
  layers:
    - name: runtime-0
      rtds_layer:
        rtds_config:
          api_config_source:
            api_type: GRPC
            grpc_services:
              envoy_grpc:
                cluster_name: xds_cluster
        name: runtime-0
    - name: runtime-1
      rtds_layer:
        rtds_config:
          api_config_source:
            api_type: GRPC
            grpc_services:
              envoy_grpc:
                cluster_name: xds_cluster
        name: runtime-1
```
I see the following outputted by the RTDS server:
```
2019/09/17 09:04:12 stream 6 open for type.googleapis.com/envoy.service.discovery.v2.Runtime
2019/09/17 09:04:12 stream 7 open for type.googleapis.com/envoy.service.discovery.v2.Runtime
2019/09/17 09:04:12 respond type.googleapis.com/envoy.service.discovery.v2.Runtime[runtime-0] version "" with version "v0"
2019/09/17 09:04:12 respond type.googleapis.com/envoy.service.discovery.v2.Runtime[runtime-1] version "" with version "v0"
2019/09/17 09:04:12 open watch 7 for type.googleapis.com/envoy.service.discovery.v2.Runtime[runtime-0] from nodeID "test-id", version "v0"
2019/09/17 09:04:12 open watch 8 for type.googleapis.com/envoy.service.discovery.v2.Runtime[runtime-1] from nodeID "test-id", version "v0"
```

File: pkg/cache/resource.go (hunk 3 above)
Comment (ivanhan) [reply pkg/cache/resource.go]: fixed!

File: pkg/test/resource/resource.go (hunk 5 above)
Comment (ivanhan) [reply pkg/test/resource/resource.go]: Addressed! Added two dummy fields :) Didn't felt it necessary to expose the fields on the function to keep the code simple, hope that's OK

**Comment by ivanhan** [comment]
Incorporated feedback, fixed README, added RTDS to ADS and REST!

File: sample/bootstrap-ads.yaml [hunk 6]
```diff
@@ -28,3 +28,14 @@ static_resources:
         port_value: 18000
     http2_protocol_options: {}
     name: xds_cluster
+layered_runtime:
+  layers:
+    - name: runtime-0
+      rtds_layer:
+        rtds_config:
+          api_config_source:
```
Comment (kyessenov) [inline sample/bootstrap-ads.yaml]: Shouldn't this be something like `ads: {}` instead? Can you also bump up to two layers in ADS only if possible?

Comment (ivanhan) [reply sample/bootstrap-ads.yaml]: Oops!!! checking! My bad

File: sample/bootstrap-xds.yaml (hunk 2 above)
Comment (ivanhan) [reply sample/bootstrap-xds.yaml]: Oohhh!!! It looks like ADS does support retrieving multiple layers, as running ADS with 2 RTDS layers, I see:

2019/09/17 09:29:25 respond type.googleapis.com/envoy.service.discovery.v2.Runtime[runtime-1 runtime-0] version "v1" with version "v2"
2019/09/17 09:29:25 open watch 14 for type.googleapis.com/envoy.service.discovery.v2.Runtime[runtime-1 runtime-0] from nodeID "test-id", version "v2"

File: sample/bootstrap-ads.yaml (hunk 6 above)
Comment (ivanhan) [reply sample/bootstrap-ads.yaml]: Addressed! Sorry it took so long, got pulled into something

File: build/integration.sh [hunk 7]
```diff
@@ -11,7 +11,10 @@ set -o pipefail
 # Management server type. Valid values are "ads", "xds", "rest"
 XDS=${XDS:-ads}
 
-(bin/test --xds=${XDS} -debug "$@")&
+# Number of RTDS layers.
+[ "$XDS" = "ads" ] && RUNTIMES=2 || RUNTIMES=1
```
Comment (yangminzhu) [inline build/integration.sh]: nit: I feel this is a bit hard to read, could you use if else here?

Comment (ivanhan) [reply build/integration.sh]: for sure!

Comment (kyessenov) [reply build/integration.sh]: You can run with two runtimes in xDS mode I think. The missing request for the second runtime won't stop the server from responding.

File: sample/bootstrap-xds.yaml (hunk 2 above)
Comment (kyessenov) [reply sample/bootstrap-xds.yaml]: Yeah, that's how ADS is supposed to work. It should batch requests. Thanks for confirming that works in envoy.

**Comment by ivanhan** [comment]
thank you very much @yangminzhu @kyessenov !!

**Pull request merged**